File: src/core/engine/requests.ts
/**
 * Higher timeframe and other instrument reads, folded onto the chart's bars.
 *
 * A read is an expression evaluated over a **different set of bars** than the
 * chart's, whose value is then folded back onto the chart's. Two halves, and
 * they are in two files: `request-body.ts` runs the expression once per
 * requested bar, and this decides which requested bar each chart bar is allowed
 * to see. The second half is where the off-by-one lives, so it is written as
 * one comparison repeated three times rather than as three rules.
 *
 * **Bars in, buckets out.** Source bars are grouped by the bucket key of their
 * open instant (`timeframe.ts`); a bucket becomes one requested bar. The source
 * is the chart's own bars for a read of the chart's instrument, which is the
 * read `host-interface.md` 5.1 says an engine can satisfy from what it already
 * holds, and the host's answer for a read of another instrument, which it never
 * can. One fold serves both, so the two kinds of read cannot drift apart.
 *
 * **The three modes differ in one thing: how far past the chart bar the fold is
 * allowed to look.** Everything else about them is the same code.
 *
 * | Mode | Source bars it may consume | Value the chart bar takes |
 * |---|---|---|
 * | `"confirmed"` | Those opening at or before this chart bar | The last bucket that closed |
 * | `"developing"` | Those opening at or before this chart bar | The bucket this bar is inside, so far |
 * | `"lookahead"` | Those belonging to this bar's bucket or earlier | The bucket this bar is inside, in full |
 *
 * Read the middle column, because it is the repainting question written out.
 * Confirmed and developing stop at the chart bar and can be computed live;
 * lookahead reads past it, which is why it repaints on history permanently and
 * by design, and why on the newest bucket, where there is nothing past the bar
 * to read, it is the bucket so far and nothing better exists.
 *
 * **The boundary.** A bucket has begun on the first chart bar whose key is the
 * new one, because a bar's open instant is what dates it. So a confirmed read
 * steps to the closed bucket's value on the first chart bar of the next bucket,
 * holds it across every bar inside that bucket including the last, and steps
 * again on the first bar of the one after. The bar a bucket closes on still
 * reads the bucket before it: at that bar's open, the bucket it is inside had
 * not closed.
 *
 * **Warmup translates by arithmetic and not by a rule.** The expression runs on
 * requested bars, so its own warmup is counted in requested bars and its absence
 * folds onto the chart like any other value: a twenty period average of daily
 * closes is absent until twenty daily bars exist, which on a five minute chart
 * is about a month in. The `warmup` field of the compiled program is what a host
 * extends the fetch range backwards by, a floor and not a promise, and an engine
 * given less gets a read that is absent for longer rather than a wrong number.
 */
import type { Span } from '../span/index.js';
import type { HostBar } from './bars.js';
import type { Clock, EngineLimits } from './budget.js';
import type { EngineHost } from './host.js';
import type { ResolvedInput } from './inputs.js';
import type { HostFacts, ManifestEntry } from './library/index.js';
import type { Registers } from './registers.js';
import { RequestBody, bodyProgram } from './request-body.js';
import { askHost, reasonFor } from './request-plan.js';
import type { RequestPlan } from './request-plan.js';
import { bucketKeyOf } from './timeframe.js';
import type { CompiledProgram, Request } from './types.js';
import { ABSENT } from './values/index.js';
import type { Heap, Value } from './values/index.js';

export interface RequestParts {
  readonly program: CompiledProgram;
  /** Every read in the program, nested ones included, keyed by its own id. */
  readonly plans: ReadonlyMap<number, RequestPlan>;
  readonly host: EngineHost;
  readonly limits: EngineLimits;
  readonly clock: Clock | undefined;
  readonly library: readonly ManifestEntry[];
  readonly inputs: readonly ResolvedInput[];
  /** The facts a body inherits, with its own identity written over three of them. */
  readonly facts: HostFacts;
  /**
   * The heap of the machine these reads land in, read when a bar asks for it.
   *
   * A function rather than the heap, because a read written inside another lands
   * in the enclosing body's heap and that body does not exist until its own
   * reads have been built.
   */
  heapOf(): Heap;
  spanAt(line: number, column: number): Span;
}

/** Where the fold of one read stands, and what a re-executed bar restores. */
interface Mark {
  readonly cursor: number;
  readonly openKey: number | undefined;
  readonly openFrom: number;
}

interface Read {
  readonly plan: RequestPlan;
  readonly register: number;
  readonly body: RequestBody;
  /** The host's bars for a `symbol` read, or nothing when the fold is the chart's. */
  readonly source: readonly HostBar[] | undefined;
  readonly ready: boolean;
  readonly reason: string;
  /** How far into the source the fold has read. */
  cursor: number;
  openKey: number | undefined;
  /** Where in the source the bucket now forming began. */
  openFrom: number;
  open: HostBar | undefined;
  /** The value of the most recently closed bucket, which a confirmed read takes. */
  closedValue: Value;
  /** The highest bucket key ever closed, so a re-executed bar closes none twice. */
  highestClosed: number | undefined;
  mark: Mark;
}

const START: Mark = { cursor: 0, openKey: undefined, openFrom: 0 };

export class RequestSet {
  private readonly reads: Read[] = [];
  private readonly byId = new Map<number, Read>();
  private readonly parts: RequestParts;

  constructor(parts: RequestParts, requests: readonly Request[]) {
    this.parts = parts;
    for (const request of requests) {
      const plan = parts.plans.get(request.id);
      if (plan === undefined) continue;
      const read = build(parts, request, plan);
      this.reads.push(read);
      this.byId.set(request.id, read);
    }
  }

  get empty(): boolean {
    return this.reads.length === 0;
  }

  /**
   * Step 4: each read's value for this bar, written into its `"request"`
   * register.
   *
   * `fresh` says whether this is the first execution of this bar. A bar that is
   * executing again restores the fold to where it stood when the bar began and
   * rebuilds the bucket now forming from the source as it now stands, which is
   * what makes a moving bar idempotent one level down: the newest chart bar may
   * have been revised since the last execution, and the bucket it is inside has
   * to be rebuilt from the revision rather than from the reading it replaced.
   */
  fill(registers: Registers, bars: readonly HostBar[], index: number, fresh: boolean): void {
    const heap = this.parts.heapOf();
    for (const read of this.reads) {
      if (fresh) {
        read.mark = { cursor: read.cursor, openKey: read.openKey, openFrom: read.openFrom };
      } else {
        restore(read, bars);
      }
      // The body's value is the body's, in the body's heap. `carry` is what
      // makes it a value of the machine that is about to read it.
      registers.set(read.register, read.body.carry(valueOf(read, bars, index), heap));
    }
  }

  /** `req.isReady(read)`: whether the host has answered this one yet. */
  answered(id: Value): Value {
    return typeof id === 'number' ? this.byId.get(id)?.ready === true : false;
  }

  /** `req.error(read)`: the reason it failed, or the empty string when none has. */
  failure(id: Value): Value {
    return typeof id === 'number' ? this.byId.get(id)?.reason ?? '' : '';
  }
}

function build(parts: RequestParts, request: Request, plan: RequestPlan): Read {
  const view = bodyProgram(parts.program, request);
  const answer = askHost(plan, parts.host);
  const refused = answer !== undefined && 'refused' in answer ? answer.refused : undefined;
  let body: RequestBody | undefined;

  // The facts a body reads are the chart's, except the three that identify an
  // instrument: inside a read, those are the instrument and the interval the
  // read asked for. The two request calls answer about the reads written inside
  // this body, which is why the set below is built before the body is.
  let nested: RequestSet | undefined;
  const facts: HostFacts = {
    ...parts.facts,
    symbol: () => plan.query.symbol ?? parts.facts.symbol(),
    exchange: () => plan.query.exchange ?? parts.facts.exchange(),
    interval: () => plan.query.timeframe,
    requestReady: (id: Value) => nested?.answered(id) ?? false,
    requestError: (id: Value) => nested?.failure(id) ?? '',
  };
  nested = new RequestSet(
    // A read inside this one lands in this body's heap, which is built below.
    { ...parts, program: view, facts, heapOf: () => (body as RequestBody).objects },
    request.body.requests,
  );

  body = new RequestBody({
    program: view,
    limits: parts.limits,
    clock: parts.clock,
    host: facts,
    library: parts.library,
    inputs: request.body.inputs.map((one) => ({
      series: one.series,
      value: parts.inputs.find((input) => input.key === one.input)?.value ?? null,
    })),
    nested,
    spanAt: parts.spanAt,
  });

  return {
    plan,
    register: request.series,
    body,
    source: answer !== undefined && 'bars' in answer ? answer.bars : undefined,
    // A fold of the chart's own bars needs nothing from the host, so it is
    // answered the moment it is planned. A read the host is still fetching is
    // not: the read is absent, `req.isReady` is false, and the study keeps
    // drawing everything that does not depend on it.
    ready: answer === undefined ? plan.query.read === 'timeframe' : 'bars' in answer,
    reason: refused === undefined ? '' : reasonFor(refused, plan.query),
    cursor: 0,
    openKey: undefined,
    openFrom: 0,
    open: undefined,
    closedValue: ABSENT,
    highestClosed: undefined,
    mark: START,
  };
}

/** The read's value for the chart bar at `index`, and the fold that gets there. */
function valueOf(read: Read, bars: readonly HostBar[], index: number): Value {
  if (!read.ready) return ABSENT;
  const bar = bars[index];
  const at = bar?.time ?? null;
  if (at === null) return ABSENT;
  const key = keyOf(read, at);
  if (key === undefined) return ABSENT;

  const source = read.source ?? bars;
  const lookahead = read.plan.request.mode === 'lookahead';
  for (;;) {
    const next = source[read.cursor];
    if (next === undefined) break;
    const found = next.time === null ? undefined : keyOf(read, next.time);
    if (found === undefined) {
      // A source bar with no time, or none this calendar can date, belongs to
      // no bucket. It is stepped over rather than folded into whichever bucket
      // happens to be open.
      read.cursor += 1;
      continue;
    }
    // The one line the three modes differ in. A lookahead read may read to the
    // end of the bucket the chart bar is inside; the other two stop at the bar.
    if (lookahead ? found > key : (next.time ?? 0) > at) break;
    read.cursor += 1;
    feed(read, next, found);
  }

  if (read.plan.request.mode === 'confirmed') return read.closedValue;
  if (read.openKey !== key || read.open === undefined) return ABSENT;
  return read.body.peek(read.open);
}

function keyOf(read: Read, instant: number): number | undefined {
  return bucketKeyOf(instant, read.plan.timeframe, read.plan.zone);
}

/**
 * One source bar folded in.
 *
 * A bar whose key is the open bucket's extends it. A bar whose key is new
 * closes the open bucket, which is the only place a requested bar ever settles:
 * a bucket is closed by something later, never by the clock. The guard on
 * `highestClosed` is what makes that idempotent, because a chart bar that
 * closed a bucket on its first execution must not close it again on its second.
 */
function feed(read: Read, bar: HostBar, key: number): void {
  if (read.openKey === key && read.open !== undefined) {
    read.open = extend(read.open, bar);
    return;
  }
  if (read.openKey !== undefined && read.open !== undefined) {
    if (read.highestClosed === undefined || read.openKey > read.highestClosed) {
      read.closedValue = read.body.settle(read.open);
      read.highestClosed = read.openKey;
    }
  }
  read.openKey = key;
  read.openFrom = read.cursor - 1;
  read.open = start(bar);
}

/** A re-executed bar: the fold as it stood when the bar began. */
function restore(read: Read, bars: readonly HostBar[]): void {
  read.cursor = read.mark.cursor;
  read.openKey = read.mark.openKey;
  read.openFrom = read.mark.openFrom;
  read.open = undefined;
  if (read.openKey === undefined) return;
  const source = read.source ?? bars;
  for (let at = read.openFrom; at < read.cursor; at += 1) {
    const bar = source[at];
    if (bar === undefined || bar.time === null) continue;
    if (keyOf(read, bar.time) !== read.openKey) continue;
    read.open = read.open === undefined ? start(bar) : extend(read.open, bar);
  }
}

/**
 * The first source bar of a bucket, as a bar of its own.
 *
 * The bucket's `time` is this bar's open instant rather than the instant the
 * bucket's key names. The two differ whenever trading begins after the boundary,
 * which is every session on a daily fold, and the first bar's instant is a
 * reading from the data while the boundary is a number the engine chose.
 */
function start(bar: HostBar): HostBar {
  return {
    open: bar.open,
    high: bar.high,
    low: bar.low,
    close: bar.close,
    volume: bar.volume ?? null,
    time: bar.time,
    openInterest: bar.openInterest ?? null,
  };
}

/**
 * One more source bar in the same bucket.
 *
 * **Absence propagates through every field**, which is `host-interface.md` 3.1:
 * an absent price is a hole and what it feeds turns absent. So a bucket whose
 * high is missing from one of its bars has no high, rather than the highest of
 * the bars that did report one, which would be a number with no name. Open is
 * the first bar's, close is the last bar's, and open interest is a level rather
 * than a flow so the coarser bar takes the last and never the sum.
 */
function extend(into: HostBar, bar: HostBar): HostBar {
  return {
    open: into.open,
    high: higher(into.high, bar.high),
    low: lower(into.low, bar.low),
    close: bar.close ?? null,
    volume: both(into.volume ?? null, bar.volume ?? null),
    time: into.time,
    openInterest: bar.openInterest ?? null,
  };
}

function higher(a: number | null, b: number | null | undefined): number | null {
  if (a === null || typeof b !== 'number') return null;
  return b > a ? b : a;
}

function lower(a: number | null, b: number | null | undefined): number | null {
  if (a === null || typeof b !== 'number') return null;
  return b < a ? b : a;
}

function both(a: number | null, b: number | null): number | null {
  return a === null || b === null ? null : a + b;
}

File: engine/openscript/requests.py
"""Higher timeframe and other instrument reads, folded onto the chart's bars.

A read is an expression evaluated over a **different set of bars** than the
chart's, whose value is then folded back onto the chart's. Two halves, in two
modules: ``request_body.py`` runs the expression once per requested bar, and this
decides which requested bar each chart bar is allowed to see. The second half is
where the off-by-one lives, so it is one comparison written once rather than
three rules.

**Bars in, buckets out.** Source bars are grouped by the bucket key of their open
instant (``buckets.py``); a bucket becomes one requested bar. The source is the
chart's own bars for a read of the chart's instrument, which is the read
`host-interface.md` 5.1 says an engine satisfies from what it already holds, and
the host's answer for a read of another instrument, which it never can. One fold
serves both, so the two kinds of read cannot drift apart.

**The three modes differ in one thing: how far past the chart bar the fold may
look**, which is `compiled-program.md` 2.16.2's table. ``"confirmed"`` and
``"developing"`` consume source bars opening at or before the chart bar, and
take the last bucket that closed and the bucket this bar is inside so far;
``"lookahead"`` consumes every source bar of this bar's bucket, and takes that
bucket in full. A bucket closes when a bar of a later one is folded, never by
the clock, so a confirmed read steps on the first chart bar of the next bucket.

**A re-executed chart bar restores where each fold stood when the bar began**, and
rebuilds the bucket then forming from the source as it now stands, which is
what makes a moving bar idempotent one level down: the newest chart bar may
have been revised since the last execution, and its bucket is rebuilt from the
revision rather than from the reading it replaced. ``mark`` and ``restore`` are
that position, taken into ``run.py``'s checkpoint beside the cells. What the
position does not hold is the value of a bucket that has closed and the body's
own state: a closed bucket is closed by a bar before the newest one, which no
revision touches, and a second close of it is refused below rather than undone.
"""

from dataclasses import dataclass, field, replace
from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence, Tuple

from .budget import EngineLimits
from .contracts import Bar, CallContext, Library
from .inputs import ResolvedInput
from .buckets import bucket_key
from .memory import Register
from .program import LoadedProgram
from .request_body import RequestBody, body_program, carry
from .request_plan import Answered, Plan, Provider, Refusal, ask_host, reason_for, undatable, undatable_reason
from .values import ABSENT, is_number, stored

#: Where one read's fold stands: how far into the source it has read, the key of
#: the bucket forming, and where in the source that bucket began.
Mark = Tuple[int, Optional[int], int]


@dataclass(frozen=True)
class RequestParts:
    """What every read of one machine is built from."""

    program: LoadedProgram
    #: Every read of the program, nested ones included, by its own id.
    plans: Mapping[int, Plan]
    provider: Optional[Provider]
    limits: EngineLimits
    library: Library
    inputs: Sequence[ResolvedInput]


@dataclass
class Read:
    """One read: its plan, the body, the source, and where the fold stands."""

    plan: Plan
    register: int
    body: RequestBody
    #: The host's bars for this read, or nothing when the fold is the chart's.
    source: Optional[Sequence[Bar]]
    ready: bool
    reason: str
    cursor: int = 0
    open_key: Optional[int] = None
    open_from: int = 0
    open: Optional[Bar] = None
    #: The value of the most recently closed bucket, which a confirmed read takes.
    closed_value: Any = ABSENT
    #: The highest bucket key ever closed, so a re-executed bar closes none twice.
    highest_closed: Optional[int] = field(default=None)


class RequestSet:
    """The reads one machine makes, in the order its ``requests`` table lists them."""

    def __init__(self, parts: RequestParts, requests: Sequence[Any]) -> None:
        self.reads: List[Read] = []
        self._by_id: Dict[int, Read] = {}
        for request in requests:
            plan = parts.plans.get(request["id"])
            if plan is None:
                continue
            read = _build(parts, request, plan)
            self.reads.append(read)
            self._by_id[request["id"]] = read

    @property
    def empty(self) -> bool:
        return not self.reads

    # -- the checkpoint --------------------------------------------------------

    def mark(self) -> Tuple[Mark, ...]:
        return tuple((one.cursor, one.open_key, one.open_from) for one in self.reads)

    def restore(self, mark: Sequence[Mark], bars: Sequence[Any]) -> None:
        """Each fold as it stood, and the bucket then forming rebuilt from its source."""
        for read, (cursor, open_key, open_from) in zip(self.reads, mark):
            read.cursor, read.open_key, read.open_from = cursor, open_key, open_from
            read.open = None
            if open_key is None:
                continue
            source = bars if read.source is None else read.source
            for at in range(open_from, cursor):
                bar = _at(source, at)
                if bar is None or _key(read, bar.time) != open_key:
                    continue
                read.open = _start(bar) if read.open is None else _extend(read.open, bar)

    # -- step 4 ----------------------------------------------------------------

    def fill(self, registers: List[Register], bars: Sequence[Any], index: int, context: CallContext) -> None:
        """Each read's value for the chart bar at ``index``, into its register."""
        for read in self.reads:
            registers[read.register].current = stored(carry(_value_of(read, bars, index, context)))

    # -- req.isReady and req.error ----------------------------------------------

    def answered(self, handle: Any) -> bool:
        """Whether the host has answered this read, which a read with a reason never is."""
        read = self._by_id.get(handle) if is_number(handle) else None
        return read is not None and read.ready

    def failure(self, handle: Any) -> str:
        """The reason this read will never answer, or the empty string when none has."""
        read = self._by_id.get(handle) if is_number(handle) else None
        return "" if read is None else read.reason


def _build(parts: RequestParts, request: Dict[str, Any], plan: Plan) -> Read:
    view = body_program(parts.program, request)
    answer = ask_host(plan, parts.provider)
    nested = RequestSet(replace(parts, program=view), request["body"]["requests"])
    # A setting a body reads is the enclosing program's, resolved once at load.
    settings = {one.key: one.value for one in parts.inputs if one.value is not None}
    inputs = [(one["series"], settings.get(one["input"], ABSENT)) for one in request["body"]["inputs"]]
    body = RequestBody(view, plan, parts.limits, parts.library, inputs, nested)
    # A reason is the read saying why it will never answer. The host's own words
    # win where both hold: a source that refused is the fix a user can act on first.
    if isinstance(answer, Refusal):
        why = reason_for(answer, plan.query)
    else:
        why = undatable_reason(plan.query) if undatable(plan) else ""
    # A fold of the chart's own bars needs nothing from the host, so it is
    # answered the moment it is planned; a read the host is still fetching is not.
    if answer is None:
        answered = plan.query.read == "timeframe"
    else:
        answered = isinstance(answer, Answered)
    return Read(
        plan=plan,
        register=request["series"],
        body=body,
        source=list(answer.bars) if isinstance(answer, Answered) else None,
        ready=why == "" and answered,
        reason=why,
    )


def _at(source: Sequence[Any], index: int) -> Any:
    return source[index] if 0 <= index < len(source) else None


def _key(read: Read, instant: Any) -> Optional[int]:
    return bucket_key(instant, read.plan.timeframe, read.plan.zone)


def _value_of(read: Read, bars: Sequence[Any], index: int, context: CallContext) -> Any:
    """The read's value for one chart bar, and the fold that gets there."""
    if not read.ready:
        return ABSENT
    bar = _at(bars, index)
    at = None if bar is None else bar.time
    key = None if at is None else _key(read, at)
    if key is None:
        return ABSENT
    source = bars if read.source is None else read.source
    lookahead = read.plan.request["mode"] == "lookahead"
    while True:
        after = _at(source, read.cursor)
        if after is None:
            break
        found = _key(read, after.time)
        if found is None:
            # A source bar this calendar cannot date belongs to no bucket, and is
            # stepped over rather than folded into whichever one is open.
            read.cursor += 1
            continue
        # The one line the three modes differ in.
        if (found > key) if lookahead else (after.time > at):
            break
        read.cursor += 1
        _feed(read, after, found, context)
    if read.plan.request["mode"] == "confirmed":
        return read.closed_value
    if read.open_key != key or read.open is None:
        return ABSENT
    return read.body.peek(read.open, context)


def _feed(read: Read, bar: Bar, key: int, context: CallContext) -> None:
    """One source bar folded in: it extends the bucket forming, or closes it and begins the next."""
    if read.open_key == key and read.open is not None:
        read.open = _extend(read.open, bar)
        return
    if read.open_key is not None and read.open is not None:
        if read.highest_closed is None or read.open_key > read.highest_closed:
            read.closed_value = read.body.settle(read.open, context)
            read.highest_closed = read.open_key
    read.open_key = key
    read.open_from = read.cursor - 1
    read.open = _start(bar)


def _price(value: Any) -> Any:
    return float(value) if is_number(value) else ABSENT


def _start(bar: Bar) -> Bar:
    """The first source bar of a bucket, as a bar of its own, dated by its own open.

    The bucket's time is this bar's open instant rather than the boundary the
    key names: the two differ whenever trading begins after the boundary, and
    the first is a reading from the data while the second is a number the
    engine chose.
    """
    return Bar(
        time=bar.time,
        open=_price(bar.open),
        high=_price(bar.high),
        low=_price(bar.low),
        close=_price(bar.close),
        volume=_price(bar.volume),
        oi=_price(bar.oi),
    )


def _extend(into: Bar, bar: Bar) -> Bar:
    """One more source bar in the same bucket.

    **Absence propagates through every field** (`host-interface.md` 3.1): a
    bucket missing one bar's high has no high, rather than the highest of the
    bars that did report one. Open is the first bar's, close the last one's, and
    open interest is a level, so the coarser bar takes the last and not the sum.
    """
    return Bar(
        time=into.time,
        open=into.open,
        high=_either(into.high, bar.high, lambda a, b: b if b > a else a),
        low=_either(into.low, bar.low, lambda a, b: b if b < a else a),
        close=_price(bar.close),
        volume=_either(into.volume, bar.volume, lambda a, b: a + b),
        oi=_price(bar.oi),
    )


def _either(held: Any, more: Any, combine: Callable[[float, float], float]) -> Any:
    if held is ABSENT or not is_number(more):
        return ABSENT
    return combine(held, float(more))


#: ``req.isReady`` and ``req.error``, by name and argument count as the manifest
#: keys them. Each takes the reads of the machine the call is in and the
#: arguments; a machine with none has answered nothing and reported no reason.
CALLS: Dict[Tuple[str, int], Callable[[Any, Sequence[Any]], Any]] = {
    ("req.isReady", 1): lambda reads, arguments: False if reads is None else reads.answered(arguments[0]),
    ("req.error", 1): lambda reads, arguments: "" if reads is None else reads.failure(arguments[0]),
}
